### PR TITLE
[air][tune] Aim logger 

### LINKED ISSUE #30537: Integration of aim metric tracker as an alternative to tensorboardx in ray tune
### Description

I suggest integrating the aim metric tracker to ray tune as an alternative to tensorboardx. 

I already have a working implementation of this inspired by https://github.com/ray-project/ray/blob/master/python/ray/tune/logger/tensorboardx.py and I am already in touch with aim: 
https://github.com/aimhubio/aim/issues/2369

If this feature is considered useful I'm happy to implement it.

### Use case

Tensorboardx has some donwsides to it like poor scalability and having another open source metrics tracker as a possible alternative could be quite valuable.

### PR DESCRIPTION
**Pull request by ju2ez**
## Why are these changes needed?

The default tensorboardx logger, though very powerful has some downsides to it like poor scalability (slow with many experiments). With this PR I want to give the option for an alternative (powerful) logger in form of the aim logger. 

## Related issue number

Closes #30537 

Fixes #30674 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] proper unit testing of the written aim metrics
   - [ ] Release tests

### CODE REVIEW

File: doc/source/tune/examples/tune-aim.ipynb
Comment (angelinalg) [inline doc/source/tune/examples/tune-aim.ipynb:13]:
```suggestion
    "[Aim](https://aimstack.io) is an easy-to-use and supercharged open-source experiment tracker.\n",
```

File: doc/source/tune/api_docs/logging.rst
Comment (angelinalg) [inline doc/source/tune/api_docs/logging.rst:22]:
```suggestion
Install Aim via ``pip install aim``.
```

File: doc/source/tune/api_docs/logging.rst
Comment (angelinalg) [inline doc/source/tune/api_docs/logging.rst:23]:
```suggestion
See the :doc:`tutorial here </tune/examples/tune-aim>`
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (angelinalg) [inline doc/source/tune/examples/tune-aim.ipynb:14]:
```suggestion
    "Aim logs your training runs, enables a well-designed UI to compare them, and provides an API to query them programmatically.\n",
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (angelinalg) [inline doc/source/tune/examples/tune-aim.ipynb:24]:
```suggestion
    "The {ref}`AimCallback <tune-aim-logger>` automatically logs metrics and reports them to Tune with the Aim API.\n",
```

Comment (angelinalg) [reply doc/source/tune/examples/tune-aim.ipynb:24]: I apologize if I have changed the meaning of your sentence. Feel free to fix.

File: doc/source/tune/examples/tune-aim.ipynb
Comment (angelinalg) [inline doc/source/tune/examples/tune-aim.ipynb:34]:
```suggestion
    "The following example demonstrates how the `AimCallback` works in a sample project.\n",
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (angelinalg) [inline doc/source/tune/examples/tune-aim.ipynb:35]:
```suggestion
    "Begin by importing the necessary modules:"
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (angelinalg) [inline doc/source/tune/examples/tune-aim.ipynb:59]:
```suggestion
    "Next, define a simple `train_function` (a Tune `Trainable`) that reports a random loss to Tune.\n",
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (angelinalg) [inline doc/source/tune/examples/tune-aim.ipynb:137]:
```suggestion
    "When the script executes, a grid-search is carried out and the results are seamlessly saved to the Aim repo.\n",
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (angelinalg) [inline doc/source/tune/examples/tune-aim.ipynb:138]:
```suggestion
    "You can easily access and analyze the results with the user-friendly interface provided by the Aim UI, stored in the `ray_tune` repository."
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (angelinalg) [inline doc/source/tune/examples/tune-aim.ipynb:161]:
```suggestion
    "The next sections contain more in-depth information on the API of the Tune-Aim integration.\n",
```

File: python/ray/tune/logger/aim.py
Comment (angelinalg) [inline python/ray/tune/logger/aim.py:36]:
```suggestion
    performant and well-designed UI.
```

File: python/ray/tune/logger/aim.py
Comment (angelinalg) [inline python/ray/tune/logger/aim.py:54]:
```suggestion
    For more arguments please see the Aim documentation: https://aimstack.readthedocs.io/en/latest/refs/sdk.html
```

File: python/ray/tune/logger/aim.py
Comment (angelinalg) [inline python/ray/tune/logger/aim.py:46]:
```suggestion
        metrics (:obj:`List[str]`, optional): Specific metrics to track.
```

File: python/ray/tune/logger/aim.py
Comment (angelinalg) [inline python/ray/tune/logger/aim.py:47]:
```suggestion
            If no metric is specified, log everything that is reported.
```

File: python/ray/tune/logger/aim.py
Comment (angelinalg) [inline python/ray/tune/logger/aim.py:69]:
```suggestion
        See help(AimCallback) for more information about parameters.
```

**Comment by angelinalg** [comment]
I did a quick editing pass. Apologies if I mangled any sentences as I am still onboarding. Feel free to fix.

File: doc/source/tune/examples/tune-aim.ipynb
Comment (ju2ez) [reply doc/source/tune/examples/tune-aim.ipynb:24]: Yes, it is actually the other way around. :) We report metrics to tune and aim logs them

**Comment by ju2ez** [comment]
> I did a quick editing pass. Apologies if I mangled any sentences as I am still onboarding. Feel free to fix.

@angelinalg , welcome!
Your suggestions look good and I committed them. 

@justinvyu , maybe you can have another look over the current state

File: python/ray/tune/logger/aim.py
Comment (justinvyu) [inline python/ray/tune/logger/aim.py:129]: I got an error when using `as_multirun=True`, on the line where `context["trial"]` is being set.

```suggestion
        context = tmp_result.pop("context", {})
```

File: python/ray/tune/logger/aim.py
Comment (justinvyu) [inline python/ray/tune/logger/aim.py:123]:
```suggestion
        step = result[TIMESTEPS_TOTAL] or result[TRAINING_ITERATION]
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (justinvyu) [inline doc/source/tune/examples/tune-aim.ipynb:165]: Remove this header. (The autogenerated API ref already includes the name.)

File: doc/source/tune/examples/tune-aim.ipynb
Comment (justinvyu) [inline doc/source/tune/examples/tune-aim.ipynb:163]:
```suggestion
    "## Tune Aim Logger API\n",
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (justinvyu) [inline doc/source/tune/examples/tune-aim.ipynb:59]: Link to trainable docs.

```suggestion
    "Next, define a simple `train_function` (a Tune [`Trainable`](trainable-docs)) that reports a random loss to Tune.\n",
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (justinvyu) [inline doc/source/tune/examples/tune-aim.ipynb:100]: Nit: We can just remove this outer function and have the inside be a single cell.

File: python/ray/tune/logger/aim.py
Comment (justinvyu) [inline python/ray/tune/logger/aim.py:31]: To be consistent with the other similar callbacks, we should name it `AimLoggerCallback`.

```suggestion
class AimLoggerCallback(LoggerCallback):
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (justinvyu) [inline doc/source/tune/examples/tune-aim.ipynb:110]: We should provide one example where the default `AimCallback()` is passed in, then explain in another cell the other parameters that are possible. Something like this:

```python
    callbacks=[AimLoggerCallback()],
# ...
tuner.fit()
```

```markdown
In the example above, we used the default configuration for the `AimLoggerCallback`. 
There are a few options that can be configured as arguments to the callback. For example, 
setting `AimLoggerCallback(repo="/path/to/repo")` will log results to the Aim repo at that 
filepath, which could be useful if you have a central location where the results of multiple 
Tune experiments are stored. Relative paths to the working directory where the script is 
launched can be used as well. By default, the repo will be set to the experiment log 
directory. See [the API reference](tune-aim-logger) for more configurations.
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (justinvyu) [inline doc/source/tune/examples/tune-aim.ipynb:108]: We should set the experiment local dir and name explicitly for the example usage later.

```python
run_config=RunConfig(name="aim_example", local_dir="~/ray_results")
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (justinvyu) [inline doc/source/tune/examples/tune-aim.ipynb:138]:
```suggestion
    "You can access and analyze the results with the user-friendly interface provided by the Aim UI, stored at the default location -- the experiment log directory (in this case, it's at `~/ray_results/aim_example`)."
```

Comment (justinvyu) [reply doc/source/tune/examples/tune-aim.ipynb:138]: Also, add a cell below this talking about how to use `aim up` to launch the web UI. For example:

```markdown
Now that we have logged our results to the Aim repository, we can view it in Aim's web UI.
To do this, we first navigate to the directory where the Aim repository lives, then we use
the Aim CLI to launch the web interface.
```

```bash
$ cd ~/ray_results/aim_example
$ aim up
--------------------------------------------------------------------------
                Aim UI collects anonymous usage analytics.                
                        Read how to opt-out here:                         
    https://aimstack.readthedocs.io/en/latest/community/telemetry.html    
--------------------------------------------------------------------------
Running Aim UI on repo ...
Open http://127.0.0.1:43800
Press Ctrl+C to exit
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (justinvyu) [inline doc/source/tune/examples/tune-aim.ipynb:160]:
```suggestion
```

File: doc/source/tune/examples/tune-aim.ipynb
Comment (justinvyu) [inline doc/source/tune/examples/tune-aim.ipynb:24]:
```suggestion
    "The {ref}`AimCallback <tune-aim-logger>` automatically logs metrics that are reported to Tune by using the Aim API.\n",
```

**Review by justinvyu — CHANGES_REQUESTED** [review]
Thanks @ju2ez, this should be almost good to go, I have some suggestions and a few questions.

Biggest thing to clarify is the `as_multirun` flag. What is the purpose of this, and why would I want to set it to `False`?

A few things left to add:
- aim as a Tune dependency (for the CI to run properly)
- A test case that tests the `as_multirun=False` codepath
- Fixing some of the current tests to test the default behavior (no repo/experiment arguments passed in)

If you want to add me as a collaborator to your fork, I can help out with finishing up writing the tests once the other code changes are in. Let's try to get this merged this week!

**Comment by tmynn** [comment]
@justinvyu the decision to use the `as_multirun` flag is a matter of taste. If you want to group a series of trials into a single run, set the flag to False. If, however, you prefer to store each trial separately, set the flag to True.
From a research perspective, I believe this is valuable.

**Comment by ju2ez** [comment]
> Thanks @ju2ez, this should be almost good to go, I have some suggestions and a few questions.
> 
> Biggest thing to clarify is the `as_multirun` flag. What is the purpose of this, and why would I want to set it to `False`?
> 
> A few things left to add:
> 
>     * aim as a Tune dependency (for the CI to run properly)
> 
>     * A test case that tests the `as_multirun=False` codepath
> 
>     * Fixing some of the current tests to test the default behavior (no repo/experiment arguments passed in)
> 
> 
> If you want to add me as a collaborator to your fork, I can help out with finishing up writing the tests once the other code changes are in. Let's try to get this merged this week!

@justinvyu 
Sounds good. 
You are added to the fork!
Let's finish this yes :-)

**Comment by justinvyu** [comment]
@tmynn How are hyperparameters tracked if multiple trials correspond to a single Run? Is this useful when you're doing multiple samples of the same hyperparameter configuration - and aim will create mean/variance bars for your plots?

File: python/ray/tune/logger/aim.py
Comment (justinvyu) [reply python/ray/tune/logger/aim.py]: We are removing `as_multirun`. Always create a run per trial.

File: python/ray/tune/logger/__init__.py
Comment (richardliaw) [inline python/ray/tune/logger/__init__.py:11]: We shouldnt import this here; it should be done lazily

File: python/ray/tune/logger/aim.py
Comment (richardliaw) [inline python/ray/tune/logger/aim.py:50]: I think we need to remove the asterisks for the rendering to work well

**Review by richardliaw — CHANGES_REQUESTED** [review]
mostly looks good!

File: python/ray/tune/logger/__init__.py
Comment (justinvyu) [reply python/ray/tune/logger/__init__.py:11]: I think it's okay to import here - we have a try/catch for importing the aim module stuff. Will only error out when instantiating an `AimLoggerCallback`.

File: python/ray/tune/logger/aim.py
Comment (justinvyu) [reply python/ray/tune/logger/aim.py:50]: Updated.

**Comment by ju2ez** [comment]
@justinvyu @richardliaw what's left here to do?

**Comment by justinvyu** [comment]
> @justinvyu @richardliaw what's left here to do?

Hey @ju2ez, we're currently waiting on a patch release on aim's side to fix its `requirements.txt`. Currently, aim fails on our CI due to the version of a certain dependency being incompatible and erroring out. Should be on track to merge it in this week!

**Review by justinvyu — APPROVED** [review]
@richardliaw Should be good after tests pass. Could you also switch your review to approve after a last look?

File: python/ray/tune/logger/__init__.py
Comment (richardliaw) [reply python/ray/tune/logger/__init__.py:11]: This isn't consistent with what we do with the rest of the 3rd party Callbacks. Can we keep it consistent?

Comment (justinvyu) [reply python/ray/tune/logger/__init__.py:11]: I see. So TBX is not considered "3rd party" since it's default.

Comment (justinvyu) [reply python/ray/tune/logger/__init__.py:11]: Gonna remove this import then.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**
